sdks/python/apache_beam/testing/benchmarks/nexmark/nexmark_launcher.py

```diff
@@ -0,0 +1,242 @@
+      sys.exit(1)
+
+    # We use the save_main_session option because one or more DoFn's in this
+    # workflow rely on global context (e.g., a module imported at module level).
+    self.pipeline_options.view_as(SetupOptions).save_main_session = True
+
+  def generate_events(self):
+    publish_client = pubsub.Client(project=self.project)
+    topic = publish_client.topic(self.topic_name)
+    if topic.exists():
+      topic.delete()
+    topic.create()
+    sub = topic.subscription(self.subscription_name)
+    if sub.exists():
+      sub.delete()
+    sub.create()
+
+    logging.info('Generating auction events to topic %s', topic.name)
+
+    if self.args.input.startswith('gs://'):
+      from apache_beam.io.gcp.gcsfilesystem import GCSFileSystem
+      fs = GCSFileSystem(self.pipeline_options)
+      with fs.open(self.args.input) as infile:
+        for line in infile:
+          topic.publish(line)
+    else:
+      with open(self.args.input) as infile:
+        for line in infile:
+          topic.publish(line)
+
+    logging.info('Finished event generation.')
+
+    # Read from PubSub into a PCollection.
+    if self.args.subscription_name:
+      raw_events = self.pipeline | 'ReadPubSub' >> beam.io.ReadFromPubSub(
+          subscription=sub.full_name)
+    else:
+      raw_events = self.pipeline | 'ReadPubSub' >> beam.io.ReadFromPubSub(
+          topic=topic.full_name)
+
+    return raw_events
+
+  def run_query(self, query, query_error):
+    try:
+      self.pipeline = beam.Pipeline(options=self.pipeline_options)
+      raw_events = self.generate_events()
+      query.load(raw_events)
+      result = self.pipeline.run()
+      job_duration = (
+          self.pipeline_options.view_as(TestOptions).wait_until_finish_duration
+      )
+      if self.pipeline_options.view_as(StandardOptions).runner == 'DataflowRunner': # pylint: disable=line-too-long
+        result.wait_until_finish(duration=job_duration)
+        result.cancel()
+      else:
+        result.wait_until_finish()
+    except Exception as exc:
+      query_error.append(str(exc))
+      raise
+
+  def cleanup(self):
+    publish_client = pubsub.Client(project=self.project)
+    topic = publish_client.topic(self.topic_name)
+    if topic.exists():
+      topic.delete()
+    sub = topic.subscription(self.subscription_name)
+    if sub.exists():
+      sub.delete()
+
+  def run(self):
+    queries = {
+        0: query0,
+        # TODO(mariagh): Add more queries.
+    }
+
+    for i in self.args.query:
+      logging.info('Running query %d', i)
+
+      # The DirectRunner is the default runner, and it needs
+      # special handling to cancel streaming jobs.
+      launch_from_direct_runner = self.pipeline_options.view_as(
+          StandardOptions).runner in [None, 'DirectRunner']
+
+      if launch_from_direct_runner:
+        query_errors = collections.defaultdict(list)
+        command = Command(self.run_query, args=[queries[i], query_errors[i]])
```

Comment:
Maybe I am missing something. Please explain if that is the case.
`query_errors = collections.defaultdict(list)` is in the same loop as all of its uses. I do not understand using a dictionary and keying it on i, instead of just making it a list and passing that to run_query.

Reply:
I meant to define the defaultdict outside of the loop. I see now that using a list is enough.